### PR TITLE
YJIT: Move exits in gen_send_iseq to functions and use ?

### PR DESCRIPTION
**Pull request by jimmyhmiller**
Happy to bike shed on names. Just wanted to get the shape out there.

### CODE REVIEW

File: yjit/src/codegen.rs [hunk 1]
```diff
@@ -245,6 +245,16 @@ pub enum JCCKinds {
     JCC_JNA,
 }
 
+fn gen_increment_count(asm: &mut Assembler, counter: Counter) {
```
Comment (k0kubun) [inline yjit/src/codegen.rs:248]: I think this should check `if (get_option!(gen_stats))` like `gen_counter_incr` does. `gen_counter_incr` could share this implementation if it calls `gen_increment_count(asm, Counter::$counter_name)`.

File: yjit/src/codegen.rs [hunk 2]
```diff
@@ -6227,6 +6135,155 @@ fn gen_send_iseq(
...
+fn exit_if_has_rest_and_optional_and_block(asm: &mut Assembler, iseq_has_rest: bool, opt_num: i32, iseq: *const rb_iseq_t, block_arg: bool) -> Option<()> {
+    exit_if(
+        asm,
+        iseq_has_rest && opt_num != 0 && (unsafe { get_iseq_flags_has_block(iseq) } || block_arg),
+        Counter::send_iseq_has_rest_opt_and_block
+    )
+}
+
+#[must_use]
+fn exit_if_unsupported_block_arg_type(asm: &mut Assembler, block_arg_type: Option<Type>) -> Option<()> {
```
Comment (k0kubun) [inline yjit/src/codegen.rs:6259]: `gen_send_cfunc` could use this too, but is it out of scope in this PR?

File: yjit/src/codegen.rs (hunk 1 above)
Comment (jimmyhmiller) [reply yjit/src/codegen.rs:248]: Done. Should we just replace gen_counter_incr! with a non-macro version?

File: yjit/src/codegen.rs (hunk 2 above)
Comment (jimmyhmiller) [reply yjit/src/codegen.rs:6250]: I assume cfunc and iseq will probably get support for block types at different times. So I didn't extract that out into a function.

Comment (k0kubun) [reply yjit/src/codegen.rs:6250]: [This place](https://github.com/ruby/ruby/blob/165ef42b3c29414a29c3720bd353c990bb9a6638/yjit/src/codegen.rs#L5545-L5556) currently has the exact same code, so I thought you could just call `exit_if_unsupported_block_arg_type(asm, block_arg_type)?;` there, but do you mean it's gonna change soon?

Comment (k0kubun) [reply yjit/src/codegen.rs:6250]: But adding support for other block types to cfunc and iseq at different times does make sense. If sharing it makes it more difficult, it's fine to leave it as is. I wonder how you're going to do the same kind of refactoring on `gen_send_cfunc` though (the function names will conflict if it's not shared).

File: yjit/src/codegen.rs (hunk 1 above)
Comment (maximecb) [reply yjit/src/codegen.rs:248]: We could replace gen_counter_incr with the method version and use the same name as before.

File: yjit/src/codegen.rs [hunk 3]
```diff
@@ -6227,6 +6126,155 @@ fn gen_send_iseq(
     Some(EndBlock)
 }
 
+fn exit_if(asm: &mut Assembler, pred: bool, counter: Counter) -> Option<()> {
```
Comment (maximecb) [inline yjit/src/codegen.rs:6129]: Should comment to explain what this is used for. Put yourself in the shoes of someone who's never seen this codebase and maybe is relatively new to Rust.

File: yjit/src/codegen.rs (hunk 1 above)
Comment (jimmyhmiller) [reply yjit/src/codegen.rs:248]: @maximecb Do you want me to do this as part of this PR? Or separately?

Comment (maximecb) [reply yjit/src/codegen.rs:248]: Would be nice if it was in this PR.

Comment (jimmyhmiller) [reply yjit/src/codegen.rs]: @maximecb Done

**Comment by XrXr** [comment]
> For some reason the actions are pulling code that doesn't match what I see on the branch

PR action jobs run against a merged SHA. It's shown in the environment: `GITPULLOPTIONS: --no-tags origin refs/pull/7725/merge`.

**Comment by jimmyhmiller** [comment]
I can recreate the failures I was getting on master. Running a bisect

**Pull request merged**